pyccl/base.py

```diff
@@ -298,3 +299,332 @@ def increment(self):
 
     def reset(self):
         self.counter = 0
+
+
+def auto_assign(func, sig):
+    """Decorator to automatically assign all parameters as instance attributes.
+    This ought to be applied on constructor methods.
+
+    Arguments:
+        func (``function``):
+            Function which takes the instance as its first argument.
+            All function arguments will be assigned as attributes of the
+            instance.
+        sig (``inspect.Signature``):
+            The signature of the constructor this decorator is applied to.
+            This is needed to distinguish between parent class methods.
+    """
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        bound = sig.bind_partial(self, *args, **kwargs)
+        bound.apply_defaults()
+        bound.arguments.pop("self")
+        for name, value in bound.arguments.items():
+            setattr(self, name, value)
+        # func may now override the attributes we just set
+        func(self, *args, **kwargs)
+    return wrapper
```

Comment:
Why do we always want to do this. Won't this bloat certain objects (e.g. Pk2Ds) by making them store the input arrays both at the python and c levels?

Reply:
`auto_assign` is a helper and is only used in `CCLHalosObject`. I figured, since we already save the passed arguments of initialized objects in the `halos` subpackage as attributes, we might as well control that centrally by a purpose-made object which does that automatically.
This also helps with the `repr` of the object. Users can see what arguments where used to initialize, say, an `HMCalculator`.

Reply:
And even for `CCLHalosObject` it worries me that we're doing this. It may not be the case right now, but in the future I can see us writing profiles whose initialisation depends on a large amount of data that never gets used again. One example is my plans for a future CIB profile that takes in a bunch of infrared spectra at different redshifts and then interpolates through them to get a single spectrum. Storing all the spectra is unnecessary.

Reply:
> And even for CCLHalosObject it worries me that we're doing this.

I see. We can discuss it. I wasn't too keen on doing that either. Let's discuss it and find a better way to do this.

Reply:
OK, to summarise what we discussed on slack: you'll remove `auto_assign`